stdlib/LinearAlgebra/src/symmetric.jl

```diff
@@ -496,6 +496,121 @@ for (T, trans, real) in [(:Symmetric, :transpose, :identity), (:(Hermitian{<:Uni
     end
 end
 
+function kron(A::Hermitian{T}, B::Hermitian{S}) where {T<:Union{Real,Complex},S<:Union{Real,Complex}}
+    resultuplo = A.uplo == 'U' || B.uplo == 'U' ? :U : :L
+    R = Hermitian(Matrix{promote_op(*, T, S)}(undef, _kronsize(A, B)), resultuplo)
+    return kron!(R, A, B)
+end
+
+function kron(A::Symmetric{T}, B::Symmetric{S}) where {T,S}
+    resultuplo = A.uplo == 'U' || B.uplo == 'U' ? :U : :L
+    R = Symmetric(Matrix{promote_op(*, T, S)}(undef, _kronsize(A, B)), resultuplo)
+    return kron!(R, A, B)
+end
+
+for (T, conj, real) in [(:Symmetric, :identity, :identity), (:(Hermitian{<:Union{Real,Complex}}), :conj, :real)]
```

Comment:
The method should probably be restricted to `Hermitian{<:Union{Real,Complex}, <StridedMatrix}` and similarly for `Similar`, and a fallback implementation of the form `kron!(C.data, A.data, B.data)` may be added. This will ensure that sparse arrays (or other special array types) remain performant, if they provide a `kron!` implementation.

Reply:
Please forgive me my ignorance as I'm rather new to Julia. You are saying I should change the function declarations to
```julia
function kron(A::Hermitian{T, <:StridedMatrix}, B::Hermitian{S, <:StridedMatrix}) where {T<:Union{Real,Complex},S<:Union{Real,Complex}}
function kron(A::Symmetric{T, <:StridedMatrix}, B::Symmetric{S, <:StridedMatrix}) where {T,S}
```
in order to act only on dense matrices, not sparse ones. Fine, I get this. But then I should also give a fallback implementation, that only acts if the sparse type doesn't provide their own `kron!`? How? I thought the original function declaration was already doing this, as a more specialized function declaration for a specific type would take precedence.

Reply:
I see, the issue is that a generic Kronecker product over sparse arrays will still be vastly more efficient for Hermitian matrices than my dense version, so we don't want to capture that. Of course, such a sparse type can also implement its own specialization for Hermitian matrices, but that's a different issue.

The best I could come up with was
```julia
for T in [:Symmetric, :(Hermitian{<:Union{Real,Complex}})]
    @eval begin
        function kron!(C::$T, A::$T, B::$T)
            wrapper = Base.typename($T).wrapper
            return wrapper(kron!(C.data, A.data, B.data))
        end
    end
end
```
for the fallback and
```julia
for (T, conj, real) in [(:(Symmetric{<:Any,<:StridedMatrix}), :identity, :identity), (:(Hermitian{<:Union{Real,Complex}, <:StridedMatrix}), :conj, :real)]
    @eval begin
        function kron!(C::$T, A::$T, B::$T)
```
for the signature of the method itself. It's rather hideous. Is there a better solution?

Also, doesn't exactly the same issue applies to `dot`? Is there any package actually implementing faster versions of `dot` and `kron` for sparse arrays?

Reply:
I take my comment back, as `kron` forwarded to the data isn't the same:
```julia
julia> kron(H1, H2)
4×4 Matrix{Float64}:
 0.343295  0.436362  0.106336   0.135163
 0.436362  0.717831  0.135163   0.222349
 0.106336  0.135163  0.0616077  0.0783094
 0.135163  0.222349  0.0783094  0.128822

julia> Hermitian(kron(H1.data, H2.data))
4×4 Hermitian{Float64, Matrix{Float64}}:
 0.343295  0.436362  0.106336   0.135163
 0.436362  0.717831  0.219534   0.222349
 0.106336  0.219534  0.0616077  0.0783094
 0.135163  0.222349  0.0783094  0.128822
```
I think the current implementation is fine.

Reply:
Ah yes, of course, that would work only if `H == H.data`. The way it is is slow but at least correct. If a sparse type wants a fast version they need to implement it themselves.

Reply:
What happened to the `StridedMatrix` restriction? I think another option for dispatching on special matrix types is to use an unwrapping mechanism as in matmul and write a "symmetric kronecker kernel" or "triangular kronecker kernel", acting on the `data` fields:

```julia
function kron(A::Symmetric, B::Symmetric)
    resultuplo = A.uplo == 'U' || B.uplo == 'U' ? :U : :L
    C = kronprod_dest(A, B)
    return Symmetric(_symkron!(C, A.data, B.data, identity, identity, C.uplo, A.uplo, B.uplo), resultuplo)
end

function _symkron!(C, A, B, conj::TC, real::TR, Cuplo, Auplo, Buplo) where {TC,TR}
...
end
```

... as a proposal/sketch. What do you think? Packages could then overload `_symkron` for their storage types.

Reply:
Why would that help? Any package can define `kron!(::Hermitian{<:Union{Real,Complex}}, Hermitian{<:Union{Real,Complex}, MyArray}, ::Hermitian{<:Union{Real,Complex}, MyArray})` and that would take precedence, as @jishnub pointed out.

Reply:
The issue with that is that the method is dispatched over a type parameter, which, as we learned the lesson in matmul, is adding to package load time and latency. Moreover, you might be able to reuse that "kernel" in the kronecker product of triangular matrices, for instance. So, there is a multi-parameter challenge here: OTOH you want to dispatch on the "triangularity", OTOH you also want to dispatch on the storage type. In matmul, we do that by first unwrapping the outer layer and directing to specially called methods (not just `mul!`, but `generic_matmatmul!` and alike), which in turn are then overloaded for different storage types (dense/sparse/GPU).

Reply:
Fair enough, I can do that.

I can't use literally the same kernel for triangular matrices for two reasons. First, because they don't use `uplo`, but instead have two different types, `UpperTriangular` and `LowerTriangular`. I can invent a fake `uplo` for them and make it work, though. The second problem is that when they have mismatched `uplo` the result is not triangular, so I have to add another argument to throw an error (or is it enough to leave the error handling to `kron`, and say that anyone using `_symkron!` is on their own?)

Also, I don't understand what you meant with `conj::TC, real::TR`. Shouldn't `conj` and `real` be functions here? What are these types?

Reply:
BTW: These are just suggestions, so we might consider them for another PR.

For the triangular ones, one could have, e.g.,

```julia
function kron(A::UpperTriangular, B::UpperTriangular)
    # create result storage C
    return UpperTriangular(_symkron(C, A.data, B.data, identity, identity, 'U', 'U', 'U'))
end
```

> Also, I don't understand what you meant with conj::TC, real::TR. Shouldn't conj and real be functions here? What are these types?

Yes, they are functions, and I believe we would wish to force specialization on them. To do so, one needs to have those type annotations. In fact, we may wish to force aggressive constant propagation, so all the character handling gets compiled away.

Reply:
They are good suggestions, I added them. It turned out that I can't reuse the kernel for triangular matrices, because since I'm unwrapping them I can no longer rely on the zero elements being zero, so I have to do that manually. I've added a dedicated kernel for them, which is anyway slightly faster. I just hope there's someone actually interested in taking Kronecker products of triangular matrices.

I did deviate from your suggestion in one point: I added not only a `kron` method, but also a `kron!` method. Since `LinearAlgebra` currently exports them for the generic case, it would be weird if they are suddenly missing for some arguments.

Reply:
Hm, the triangular ones are pretty peculiar. I was roughly thinking one could reuse some code, but forgot about the triangle to-be-ignored... Feels a bit like code bloat, TBH. I'll let others chime in on whether they believe the triangular stuff is worth it. Personally, I'd leave it with the symmetric/hermitian optimization.

Reply:
It doesn't seem like others will chime in.

Reply:
Let's just do the hermitian for now

Reply:
:facepalm: I wasted a lot of time doing the triangular case only because @dkarrasch asked. I have no use for it. Next time please only ask for features you actually want.

Reply:
Well?